Controller Crashes when ServiceAccount has no Namespace

<!--
Thanks for taking time reporting a Capsule bug!

We do our best to keep it reliable and working, so don't hesitate adding
as many information as you can and keep in mind you can reach us on our
Clastix Slack workspace: https://clastix.slack.com, #capsule channel.   
-->

# Bug description

When I want to assign a ServiceAccount as Owner of a Tenant it obviously requires a namespace reference to where that serviceAccount is located. I didn't think about that the first time I deployed a tenant with a ServiceAccount as Owner reference. As a result the Controller keeps crashing. This shouldn't happen in a production environment, since when the controller is crashing you have little change to delete the tenant. So it would be nice if this misconfiguration could be handled smoother.

Corresponds to the following line: https://github.com/clastix/capsule/blob/master/controllers/tenant/rolebindings.go#L126

Sadly I am not really good in Go.

Love the project .  :heart: 

# How to reproduce

Steps to reproduce the behavior:

1. Provide the Capsule Tenant YAML definitions

```
apiVersion: capsule.clastix.io/v1beta1
kind: Tenant
metadata:
  name: bug-tenant
spec:
  owners:
    - kind: ServiceAccount
      name: some-service-account
```

2. Provide all managed Kubernetes resources


# Expected behavior

The Controller should reject the template with throwing the error, that a namespace is required for the serviceAccount.

# Logs

If applicable, please provide logs of `capsule`.

In a standard stand-alone installation of Capsule,
you'd get this by running `kubectl -n capsule-system logs deploy/capsule-controller-manager`.

``` 
{"level":"debug","ts":"2021-09-03T17:51:32.828Z","logger":"controller-runtime.webhook.webhooks","msg":"received request","webhook":"/tenants","UID":"f3e8caf7-ea56-46e5-b649-c8c152a96e34","kind":"capsule.clastix.io/v1beta1, Kind=Tenant","resource":{"group":"capsule.clastix.io","version":"v1beta1","resource":"tenants"}}
{"level":"debug","ts":"2021-09-03T17:51:32.829Z","logger":"controller-runtime.webhook.webhooks","msg":"wrote response","webhook":"/tenants","code":200,"reason":"","UID":"f3e8caf7-ea56-46e5-b649-c8c152a96e34","allowed":true}
{"level":"info","ts":"2021-09-03T17:51:32.846Z","logger":"controllers.Tenant","msg":"Ensuring all Namespaces are collected","Request.Name":"x02-seu-java"}
{"level":"info","ts":"2021-09-03T17:51:32.854Z","logger":"controllers.Tenant","msg":"Starting processing of Namespaces","Request.Name":"x02-seu-java","items":0}
{"level":"info","ts":"2021-09-03T17:51:32.855Z","logger":"controllers.Tenant","msg":"Starting processing of Limit Ranges","Request.Name":"x02-seu-java","items":1}
{"level":"info","ts":"2021-09-03T17:51:32.855Z","logger":"controllers.Tenant","msg":"Starting processing of Resource Quotas","Request.Name":"x02-seu-java","items":1}
{"level":"info","ts":"2021-09-03T17:51:32.855Z","logger":"controllers.Tenant","msg":"Desired hard requests.storage quota is 0","Request.Name":"x02-seu-java"}
{"level":"info","ts":"2021-09-03T17:51:32.855Z","logger":"controllers.Tenant","msg":"Computed requests.storage quota for the whole Tenant is 0","Request.Name":"x02-seu-java"}
{"level":"info","ts":"2021-09-03T17:51:32.855Z","logger":"controllers.Tenant","msg":"Desired hard requests.cpu quota is 1","Request.Name":"x02-seu-java"}
{"level":"info","ts":"2021-09-03T17:51:32.855Z","logger":"controllers.Tenant","msg":"Computed requests.cpu quota for the whole Tenant is 0","Request.Name":"x02-seu-java"}
{"level":"info","ts":"2021-09-03T17:51:32.855Z","logger":"controllers.Tenant","msg":"Desired hard requests.memory quota is 2Gi","Request.Name":"x02-seu-java"}
{"level":"info","ts":"2021-09-03T17:51:32.855Z","logger":"controllers.Tenant","msg":"Computed requests.memory quota for the whole Tenant is 0","Request.Name":"x02-seu-java"}
{"level":"info","ts":"2021-09-03T17:51:32.855Z","logger":"controllers.Tenant","msg":"Ensuring additional RoleBindings for owner","Request.Name":"x02-seu-java"}
{"level":"info","ts":"2021-09-03T17:51:32.855Z","logger":"controllers.Tenant","msg":"Ensuring RoleBinding for owner","Request.Name":"x02-seu-java"}
panic: runtime error: index out of range [-1]

goroutine 712 [running]:
github.com/clastix/capsule/controllers/tenant.(*Manager).ownerRoleBinding(0xc0009971c0, 0xc003e1ca00, 0x1e, 0x0)
	/workspace/controllers/tenant/rolebindings.go:129 +0xf07
github.com/clastix/capsule/controllers/tenant.Manager.Reconcile(0x1ad3dd8, 0xc0002f34a0, 0x1aca880, 0xc0004981b0, 0xc0004a1500, 0x1abf390, 0xc000812080, 0x1ac1d78, 0xc0009a1320, 0x0, ...)
	/workspace/controllers/tenant/manager.go:102 +0x5f6
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc0001ea1e0, 0x1ac1cd0, 0xc0007f5a80, 0x1737340, 0xc003f16900)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.9.5/pkg/internal/controller/controller.go:298 +0x30d
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc0001ea1e0, 0x1ac1cd0, 0xc0007f5a80, 0x1969400)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.9.5/pkg/internal/controller/controller.go:253 +0x205
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2(0xc0004a2150, 0xc0001ea1e0, 0x1ac1cd0, 0xc0007f5a80)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.9.5/pkg/internal/controller/controller.go:214 +0x6b
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.9.5/pkg/internal/controller/controller.go:210 +0x425
```



# Additional context

- Capsule version: (`capsule --version`)
- Helm Chart version: (`helm list -n capsule-system`) 0.1.0
- Kubernetes version: (`kubectl version`) v1.21.1

--- a/main.go
+++ b/main.go
@@ -154,7 +154,7 @@ func main() {
 		route.PVC(pvc.Handler()),
 		route.Service(service.Handler()),
 		route.NetworkPolicy(utils.InCapsuleGroups(cfg, networkpolicy.Handler())),
-		route.Tenant(tenant.NameHandler(), tenant.IngressClassRegexHandler(), tenant.StorageClassRegexHandler(), tenant.ContainerRegistryRegexHandler(), tenant.HostnameRegexHandler(), tenant.FreezedEmitter()),
+		route.Tenant(tenant.NameHandler(), tenant.IngressClassRegexHandler(), tenant.StorageClassRegexHandler(), tenant.ContainerRegistryRegexHandler(), tenant.HostnameRegexHandler(), tenant.FreezedEmitter(), tenant.ServiceAccountNameHandler()),
 		route.OwnerReference(utils.InCapsuleGroups(cfg, ownerreference.Handler(cfg))),
 		route.Cordoning(tenant.CordoningHandler(cfg)),
 	)

--- a/pkg/webhook/tenant/serviceaccount_format.go
+++ b/pkg/webhook/tenant/serviceaccount_format.go
@@ -1,0 +1,66 @@
+// Copyright 2020-2021 Clastix Labs
+// SPDX-License-Identifier: Apache-2.0
+
+package tenant
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+
+	"k8s.io/client-go/tools/record"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	capsulev1beta1 "github.com/clastix/capsule/api/v1beta1"
+	capsulewebhook "github.com/clastix/capsule/pkg/webhook"
+	"github.com/clastix/capsule/pkg/webhook/utils"
+)
+
+type saNameHandler struct {
+}
+
+func ServiceAccountNameHandler() capsulewebhook.Handler {
+	return &saNameHandler{}
+}
+
+func (h *saNameHandler) validateServiceAccountName(req admission.Request, decoder *admission.Decoder) *admission.Response {
+	tenant := &capsulev1beta1.Tenant{}
+	if err := decoder.Decode(req, tenant); err != nil {
+		return utils.ErroredResponse(err)
+	}
+
+	compiler := regexp.MustCompile(`^.*:.*:.*(:.*)?$`)
+
+	for _, owner := range tenant.Spec.Owners {
+		if owner.Kind != "ServiceAccount" {
+			continue
+		}
+
+		if !compiler.MatchString(owner.Name) {
+			response := admission.Denied(fmt.Sprintf("owner name %s is not a valid Service Account name ", owner.Name))
+
+			return &response
+		}
+	}
+
+	return nil
+}
+
+func (h *saNameHandler) OnCreate(_ client.Client, decoder *admission.Decoder, _ record.EventRecorder) capsulewebhook.Func {
+	return func(_ context.Context, req admission.Request) *admission.Response {
+		return h.validateServiceAccountName(req, decoder)
+	}
+}
+
+func (h *saNameHandler) OnDelete(client.Client, *admission.Decoder, record.EventRecorder) capsulewebhook.Func {
+	return func(context.Context, admission.Request) *admission.Response {
+		return nil
+	}
+}
+
+func (h *saNameHandler) OnUpdate(_ client.Client, decoder *admission.Decoder, _ record.EventRecorder) capsulewebhook.Func {
+	return func(_ context.Context, req admission.Request) *admission.Response {
+		return h.validateServiceAccountName(req, decoder)
+	}
+}

